[BOAT] Fix logger so bulk deletes and large message deletes still get logged

BenSegal855 opened the pull request:
This PR fixes two bugs in the logger module.
1. The condition that checked if a bulk delete should be logged compared the first messages channel id to the guild id from the config <:pawa_knock_head:842001776819306526>. 
1. Messages longer than 1700 characters are now uploaded to hastebin.powercord.dev to ensure the message sent in the logs is under 2000 characters.

BenSegal855 committed:
fix: bulk delete log condition

BenSegal855 committed:
feat: log upload long messages to hastebin

cyyynthia commented on packages/boat/src/modules/mod/logger.ts:
```diff
@@ -51,6 +51,17 @@ function format (template: string, message: Message<GuildTextableChannel>): stri
     ? `\nReason: ${deleteMeta.get(message.id)}`
     : ''
 
+  const tooLong = message.content.length > 1700
+  let hasteBin = ''
+
+  if (tooLong) {
+    const res = await fetch('https://haste.powercord.dev/documents', {
+      method: 'POST',
+      body: stringifyDiscordMessage(message),
+    }).then((r) => r.json())
+    hasteBin = `<https://haste.powercord.dev/${res.key}.txt>`
+  }
+
```
```suggestion
  let msg
  if (message.content.length > 1700) {
    const res = await fetch('https://haste.powercord.dev/documents', {
      method: 'POST',
      body: stringifyDiscordMessage(message),
    }).then((r) => r.json())
    msg = `*Message too long* <https://haste.powercord.dev/${res.key}.txt>`
  } else {
    msg = stringifyDiscordMessage(message).replace(/`/g, `\`${ZWS}`) || '*No contents*'
  }

```

cyyynthia replied:
> *Ben in powercord-core:*
> I believe that will put the link in the codeblock and make it unclickable. I can try to work out a cleaner way to do it with the link sill being outside the codeblock but I think it would still need a flag.

That's actually true; to make it cleaner `attachments` could be renamed `extra` and declared with a let, and then append the hastebin url if necessary

BenSegal855 replied:
I'll make that change.

cyyynthia commented on packages/boat/src/modules/mod/logger.ts:
```diff
@@ -61,7 +72,10 @@ function format (template: string, message: Message<GuildTextableChannel>): stri
     .replace(/\$time/g, new Date(message.timestamp).toUTCString())
     .replace(/\$duration/g, prettyPrintTimeSpan(Date.now() - message.timestamp))
-    .replace(/\$message/g, stringifyDiscordMessage(message).replace(/`/g, `\`${ZWS}`) || '*No contents*')}\n${attachments}`
+    .replace(/\$message/g, tooLong
+      ? '*Message too long*'
+      : stringifyDiscordMessage(message).replace(/`/g, `\`${ZWS}`)
+      || '*No contents*')}${hasteBin}\n${attachments}`
```
```suggestion
    .replace(/\$message/g, `${msg}\n${attachments}`)}`
```

cyyynthia reviewed (changes requested):
Not a big fan of the current way the message is composed; double check my suggestions as the formatting may be a bit messed up

BenSegal855 committed:
tweak: Cleaner handeling for long messages

BenSegal855 committed:
fix: bulk deletes are only uploaded once

BenSegal855 committed:
Git was weird and didnt commit everything

cyyynthia commented on packages/boat/src/modules/mod/logger.ts:
```diff
@@ -72,10 +72,10 @@ async function format (template: string, message: Message<GuildTextableChannel>)
     .replace(/\$time/g, new Date(message.timestamp).toUTCString())
     .replace(/\$duration/g, prettyPrintTimeSpan(Date.now() - message.timestamp))
-    .replace(/\$message/g, tooLong
+    .replace(/\$message/g, !bulk && cleanContent.length > 1700
```
The length measurement can be fooled as we append additional characters in addition to the <= 1700 characters we check (zws after `), which could be problematic

cyyynthia replied:
This problem hasn't been addressed

cyyynthia commented on packages/boat/src/modules/mod/logger.ts:
```diff
@@ -94,12 +94,11 @@ async function messageDeleteBulk (this: CommandClient, msgs: Array<Message<Guild
-        ? format(LIST_TEMPLATE, msg)
-        : `A message in #${msg.channel.name} that was not cached`
-  ).join('\n\n')
+  const list = (await Promise.all(msgs.map(
+    (msg) => 'author' in msg
+      ? format(LIST_TEMPLATE, msg, true)
+      : `A message in #${msg.channel.name} that was not cached`
+  ))).join('\n\n')
```
```suggestion
  const list = []
  for (const msg of msgs) {
    list.push('author' in msg
      ? await format(LIST_TEMPLATE, msg, true)
      : `A message in #${msg.channel.name} that was not cached`)
  }
```
+ necessary changes in other places now that list is an array

BenSegal855 replied:
Any reason not to use Array.map like the below snippet and then handle the processes elsewhere?
```js
const list = msgs.map((msg) =>
    'author' in msg
      ? format(LIST_TEMPLATE, msg, true)
      : `A message in #${msg.channel.name} that was not cached`)
// Code that handles promises.
```

cyyynthia replied:
A bit less clean imo; actually forgot the await in front of format in my suggestion, that's updated

BenSegal855 committed:
tweak: use array for bulk message descriptions

cyyynthia commented on packages/boat/src/modules/mod/logger.ts:
```diff
@@ -94,13 +101,15 @@ async function messageDeleteBulk (this: CommandClient, msgs: Array<Message<Guild
-  ))).join('\n\n')
+  const list = []
+  for (const msg of msgs) {
+    const channelName = this.guilds.get(config.discord.ids.serverId)?.channels.get(msg.channel.id)?.name || msg.channel.id
```
Bulk deletes can only include messages from a single channel. Fetching for every message the channel is not a good idea.

BenSegal855 committed:
tweak: only fetch channel name once

BenSegal855 committed:
Merge branch 'my-awesome-branch' into boat/fix/logger

BenSegal855 committed:
fix: dont use markdown bulk deletes

BenSegal855 committed:
tweak: add zws before doing other stuff

cyyynthia commented on packages/boat/src/modules/mod/logger.ts:
```diff
@@ -42,15 +49,30 @@ Timestamp: $time ($duration ago)
+  if (!bulk && cleanContent.length > 1700) {
+    const res = await fetch('https://haste.powercord.dev/documents', {
+      method: 'POST',
+      body: cleanContent,
```
only the output of `stringifyDiscordMessage` is needed for here, replacing the markdown codeblocks is unnecessary

BenSegal855 replied:
While escaping backticks isn't necessary here, not adding the zws's would require another call to `stringifyDiscordMessage()` while keeping them escaped requires no additional complexity and has no visible effect in the result.

cyyynthia replied:
Addressed in fa39bc0

cyyynthia committed:
fix: don't upload escaped version to haste

Resolving blocker to get the PR merged
This change was requested on August 8, and wasn't completed 3 and a half months later...

cyyynthia reviewed (approved):
Made the appropriated changes to unblock the PR
